### PR TITLE
MARKET-2094 [Category Chart] Add the "Show Legend" checkbox to the settings.

### PR DESCRIPTION
**Pull request by Fialkaksu**
Removed chartLegendVisibility prop as redundant, updated readme
[MARKET-2094](https://bugs.backendless.com/browse/MARKET-2094)

### CODE REVIEW

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-category-chart/src/index.js [hunk 1]
```diff
@@ -43,104 +35,83 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+  options,
+  legendVisibility,
+  titleVisibility,
+  title,
+  titleFontSize,
+  yGridLineVisibility,
+  xGridLineVisibility,
+  gridLinesColor,
+  gridLinesWidth,
+}) => useMemo(() => ({
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:84]:
```suggestion
const useChartConfig = ({
  options,  legendVisibility,  titleVisibility,  title,  titleFontSize,  yGridLineVisibility,
  xGridLineVisibility,  gridLinesColor,  gridLinesWidth,
}) => useMemo(() => ({
```

File: components/bl-category-chart/src/index.js [hunk 2]
```diff
@@ -43,104 +17,166 @@ export default function CategoryChartComponent({ component, elRef }) {
...
     if (!display) {
-      return null;
+      return;
+    }
+
+    let requireRerender = false;
+
+    const optionsConfig = { ...options };
+
+    if (isDataChange(chartConfig, optionsConfig)) {
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:80]: isDataChange -> isDataChanged

File: components/bl-category-chart/src/index.js [hunk 3]
```diff
@@ -43,104 +17,166 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+
+    if (chartInstance.current && requireRerender) {
+      chartInstance.current.data = { labels, datasets };
+      chartInstance.current.update();
+    } else {
+      chartInstance.current = new Chart(chartRef.current, {
+        type,
+        data: { labels, datasets },
+        options: chartConfig,
+      });
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:135]: Should it be in separate useEffect?

```
useEffect(()=> {
  chartInstance.current = new Chart(chartRef.current, {
    type,
    data: { labels, datasets },
    options: chartConfig,
  });
  
  return () => {
    chartInstance.current.destroy();
    chartInstance.current = null;
  }
}, [])

File: components/bl-category-chart/src/index.js [hunk 4]
```diff
@@ -43,104 +17,166 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+    return false;
+  }
+
+  if (typeof prevData !== typeof nextData) {
+    return true;
+  }
+
+  if (prevData === null || prevData === undefined || nextData === null || nextData === undefined) {
+    return false;
+  }
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:162]: Why do we need it?

File: components/bl-category-chart/src/index.js [hunk 5]
```diff
@@ -43,104 +17,166 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+    let requireRerender = false;
+
+    const optionsConfig = { ...options };
+
+    if (isDataChange(chartConfig, optionsConfig)) {
+      requireRerender = true;
+      Object.assign(chartConfig, optionsConfig);
+    }
+
+    const scaleConfig = {
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:85]: scaleConfig and optionsConfig - it's not configs.
Use word `config` only for main config object

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-category-chart/src/index.js (hunk 3 above)
Comment (Fialkaksu) [reply components/bl-category-chart/src/index.js:135]: They can be separated, but in the previous version we tried to avoid this.

Comment (v-excelsior) [reply components/bl-category-chart/src/index.js:135]: Avoid what?

File: components/bl-category-chart/src/index.js [hunk 6]
```diff
@@ -43,104 +17,162 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+        cn('bl-customComponent-categoryChart', classList, { 'bl-customComponent-categoryChart--disabled': disabled })
       }>
-      <canvas ref={ chartRef } style={{ backgroundColor: chartBackgroundColor, width: "100%", height: "100%" }} />
+      <canvas ref={ chartRef } style={{ backgroundColor, width: '100%', height: '100%' }}/>
     </div>
   );
 }
 
-const useChart = (chartRef, props) => {
+function useChartLogic(component) {
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:27]: useChart

File: components/bl-category-chart/src/index.js [hunk 7]
```diff
@@ -43,104 +17,162 @@ export default function CategoryChartComponent({ component, elRef }) {
...
     if (!display) {
-      return null;
+      return;
+    }
+
+    let requireRerender = false;
+
+    const additionalOptions = { ...options };
+
+    if (isDataChanged(chartConfig, additionalOptions)) {
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:80]: It's always true here, isn't it?

File: components/bl-category-chart/src/index.js [hunk 8]
```diff
@@ -43,104 +17,162 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+    };
+
+    if (isDataChanged(chartConfig.plugins, plugins)) {
+      requireRerender = true;
+      Object.assign(chartConfig.plugins.legend, plugins.legend);
+      Object.assign(chartConfig.plugins.title, plugins.title);
+    }
+
+    if (chartInstance.current && requireRerender) {
+      chartInstance.current.data = { labels, datasets };
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:128]: Why we update data if changed something else? For example legend visibility.

File: components/bl-category-chart/src/index.js [hunk 9]
```diff
@@ -43,104 +17,162 @@ export default function CategoryChartComponent({ component, elRef }) {
...
-  const chartInstance = useRef();
+    options, yGridLineVisibility, xGridLineVisibility, gridLinesColor, gridLinesWidth,
+    legendVisibility, titleVisibility, title, titleFontSize,
+    display, type, labels, datasets
+  } = component;
+
+  const chartRef = useRef(null);
+  const chartInstance = useRef(null);
+
+  const chartConfig = {
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:37]: looks like it should be `chartOptions`

File: components/bl-category-chart/src/index.js [hunk 10]
```diff
@@ -43,104 +17,162 @@ export default function CategoryChartComponent({ component, elRef }) {
...
 
   useEffect(() => {
     if (!display) {
-      return null;
+      return;
+    }
+
+    let requireRerender = false;
+
+    const additionalOptions = { ...options };
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:78]: Why additional?

As for me we should try to compare here prevOptions(chartInstance.options or something that comes from chart ref) and nextOptions(settings shaped in `options` object on dependency change)

But now i see that we compare `chartConfig`(new object) with additionalOptions(?), scales(new object), plugins(new object). Why we do this?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-category-chart/src/index.js [hunk 11]
```diff
@@ -43,104 +17,108 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+      },
     });
 
-    return () => chartInstance.current.destroy();
-  }, [display]);
+    setChartInstance(newChartInstance);
+
+    return () => {
+      newChartInstance.destroy();
+    };
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:88]: one line

File: components/bl-category-chart/src/index.js [hunk 12]
```diff
@@ -43,104 +17,108 @@ export default function CategoryChartComponent({ component, elRef }) {
...
+        lineWidth: gridLinesWidth,
+      },
+    },
+  }), [yGridLineVisibility, xGridLineVisibility, gridLinesColor, gridLinesWidth]);
 
   useEffect(() => {
     if (!display) {
-      return null;
+      return;
     }
```
Comment (v-excelsior) [inline components/bl-category-chart/src/index.js:70]: Do we really need it?

I assume this useEffect shouldn't rely on display

Comment (Fialkaksu) [reply components/bl-category-chart/src/index.js:70]: We need this dependency because without it, we get a context getting error when creating the chart.

**Pull request merged**